custom/icds_reports/sqldata/exports/dashboard_usage.py

```diff
@@ -0,0 +1,177 @@
+
+    def convert_rs_to_matrix(self, results_queryset, main_location_type):
+        """
+
+        :param results_queryset: AwcLocation values queryset
+        :param main_location_type: parent location type to fetch only locations below that level
+        :return: location matrix similar to a dataframe and locations ids of all the locations below the given
+         location
+        """
+        location_matrix = []
+        location_ids = []
+        for result in results_queryset:
+            row_data = [result['state_id'], result['district_id'], result['block_id'], result['supervisor_id'],
+                        result['doc_id']]
+            location_matrix.append(row_data)
+        # adding only descendants to the main location
+        sub_location_types = self.location_types[self.location_types.index(main_location_type) + 1:]
+        for sub_location in sub_location_types:
+            if result[sub_location] not in location_ids:
+                location_ids.append(result[sub_location])
+        return location_matrix, location_ids
+
+    def get_location_id_string_from_location_type(self, location_type):
+        if location_type == 'awc':
+            location_type = 'doc_id'
+        else:
+            location_type = location_type + '_id'
+        return location_type
+
+    def check_if_date_in_last_week(self, date):
+        if date is None:
+            return False
+        d = datetime.datetime.strptime(date, "%Y-%m-%d")
+        now = datetime.datetime.now()
+        return (d - now).days < 7
+
+    def is_launched_from_location_id(self, location_type, location_id):
+        filter_dict = {location_type + '_id': location_id, 'is_launched': 'yes'}
+        return AggAwc.objects.filter(**filter_dict).exists()
+
+    def get_excel_data(self):
+        excel_rows = []
+        filters = [['Generated at', india_now()]]
+        headers = ['Sr.No', 'State/UT Name', 'District Name', 'Block Name', 'Username', 'Level', 'Role',
+                   'Launched?', 'Last Login', 'Logged in the last week?', 'Total', 'Child', 'Pregnant Women',
+                   'Demographics', 'System Usage', 'AWC infrastructure', 'Child Growth Monitoring List',
+                   'ICDS - CAS Monthly Register', 'AWW Performance Report', 'LS Performance Report',
+                   'Take Home Ration']
+        excel_rows.append(headers)
+        serial_count = 0
+        for user_location in self.user.get_sql_locations(self.domain):
+            # getting the location types to retrieve for this user location
+            location_type_filter = {
+                self.get_location_id_string_from_location_type(user_location.location_type_name):
+                    user_location.get_id}
+            all_awc_locations = AwcLocation.objects.filter(**location_type_filter).values(*self.required_fields)
+            # converting the result set to matrix to fetch ancestors for a given location
+            location_matrix, location_ids =\
```

Comment:
location matrix looks it contains entire hierarchy list per row per awc. Why is this location_matrix needed?
The just descendants you can easily fine by  filtering AWCLocation on appropriate aggregation level.
which intern makes me thing `convert_rs_to_matrix` method is not needed.
M I missing something very obvious?

I m happy to discuss you would like more clarity.

Reply:
i dont think we will get hiercharchy structure with get_descendants. So constructed a matrix as we need ancestors and the hierarchy as well to contruct the excel data. Hope that is clear